iwyu_output.cc

```diff
@@ -1106,11 +1106,27 @@ void ProcessForwardDeclare(OneUse* use,
   }
 
   // (A7) If --no_fwd_decls has been passed, recategorize as a full use unless
-  // the decl is in this file (in which case it must be a self-sufficient decl
-  // being used, so we can just let IWYU do its work).
-  if (!use->ignore_use() &&
-      GlobalFlags().no_fwd_decls &&
-      GetFileEntry(use->decl_loc()) != GetFileEntry(use->use_loc())) {
+  // there is a decl in this file before the use (in which case it must be a
+  // self-sufficient decl being used, so we can just let IWYU do its work).
+  if (!use->ignore_use() && GlobalFlags().no_fwd_decls) {
+    for (const Decl* curr_decl = use->decl(); curr_decl != nullptr;
+         curr_decl = curr_decl->getPreviousDecl()) {
+      if (IsBeforeInSameFile(curr_decl->getLocation(), use->use_loc()))
+        return;
+
+      // If a decl can be found in the translation unit before the use but not in
+      // this file, we set that better decl as the use's decl and suggest
+      // the header where it was found (and finally recategorize it as a full
+      // use).
+      if (IsBeforeInTranslationUnit(curr_decl->getLocation(), use->use_loc())) {
+        const NamedDecl* better_decl = DynCastFrom(curr_decl);
```

Comment:
At this point, I think we're guaranteed that `curr_decl` is a `NamedDecl`, so just an unconditional cast seems better. And if we do that, we don't need `better_decl` for anything: `use->reset_decl(cast<NamedDecl>(curr_decl));`

Reply:
So if there's a longer chain of redeclarations, this will pick the first one outside of this file? This seems a little arbitrary, but I guess it kind-of makes sense.

I'm just worried there might be arcane cases where there are multiple headers containing forward-declarations for a symbol, and we choose the wrong one at random. But maybe mappings will figure that out after the fact, in that case.